### ISSUE TITLE
termux-usb -l Command Hangs on Google Pixel 7 with Android 14

### ISSUE THREAD

**Issue by Daplugg23** [termux/termux-api#638]
**Problem description**
Problem: The termux-usb -l command is consistently hanging without listing any connected USB devices on my Google Pixel 7 running Android 14. This issue occurs even when no USB device is connected.

**Steps to reproduce**

1. Open Termux.
2. Run termux-usb -l.
3. Observe that the command hangs indefinitely without output.

**Expected behavior**
The command should either list connected USB devices or exit gracefully if none are connected.

**Additional information**

Actual Behavior: The command hangs indefinitely, requiring forceful termination of the session.

Environment:

Device: Google Pixel 7
Android Version: 14
Termux Version: 0.118.0 (F-Droid)
Termux:API Version: 0.50.1 (F-Droid) 
Termux:API Package Version 0.57

Attempts to Resolve:

Reinstalled Termux and Termux:API.
Updated and upgraded all packages within Termux.
Ensured all permissions for Termux and Termux:API are granted.

**Comment by Daplugg23** [comment]
> Seems like [termux/termux-packages#16255](https://github.com/termux/termux-packages/issues/16255)
> 
> Try this command
> 
> ```
> chmod 400 $PREFIX/libexec/termux-am/am.apk
> ```
> 
> then try to use termux-usb again

Despite following the suggested solution of applying chmod 400 $PREFIX/libexec/termux-am/am.apk, the issue with the termux-usb -l command hanging remains unresolved.

**Comment by Daplugg23** [comment]
Here's the logcat.

[logcat.txt](https://github.com/termux/termux-api/files/13367159/logcat.txt)

**Comment by agnostic-apollo** [comment]
There is no reference to `termux-usb`, you need to run `logcat` immediately after. If it was writable issue, command would exit, if it's hanging, it's a different issue. What's output for `apt show termux-am`.

**Comment by Daplugg23** [comment]
I'll try the logcat again. Here's the other information

apt show termux-am
Package: termux-am
Version: 0.6.0
Maintainer: Michal Bednarski @michalbednarski
Installed-Size: 77.8 kB
Conflicts: termux-tools (<< 0.51)
Homepage: https://github.com/termux/TermuxAm
Download-Size: 26.2 kB
APT-Manual-Installed: yes
APT-Sources: https://packages-cf.termux.dev/apt/termux-main stable/main aarch64 Packages
Description: Android Oreo-compatible am command reimplementation

**Comment by Daplugg23** [comment]
I tried the logcat again.

[logcat.txt](https://github.com/termux/termux-api/files/13367761/logcat.txt)

**Comment by agnostic-apollo** [comment]
**TL;DR: Run `adb shell "settings put global cached_apps_freezer disabled"`**

---

You won't fine the reason if `logcat` is run as termux app process, it needs to be run with `adb shell` or `su` to get system logs.

The reason is Android's cached app freezer added in Android `11` via [`bdaf16e`](https://cs.android.com/android/_/android/platform/frameworks/base/+/bdaf16e72aedf06f007c1a8b288c68ee470277d1) managed by [`CachedAppOptimizer`](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/services/core/java/com/android/server/am/CachedAppOptimizer.java). I haven't looked into it much, but apps that are cached will be frozen and intents sent to them will not be received until they are unfrozen. To unfreeze an app requires starting them with some specific reasons (check below), like starting their (launcher) activity, like Termux:API app's launcher activity. I currently don't know why this has become more of an issue on Android 14, maybe due to reducing [`DEFAULT_FREEZER_DEBOUNCE_TIMEOUT`](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/services/core/java/com/android/server/am/CachedAppOptimizer.java;l=266) from `10` minutes to `10` seconds in [`912a866d`](https://cs.android.com/android/_/android/platform/frameworks/base/+/912a866dc3bf0adb39f6eb7ee2bc2de2dfe8566b).

When an app is frozen, [`ActivityManager system_server D  freezing <pid> <package_name>`](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/services/core/java/com/android/server/am/CachedAppOptimizer.java;l=2270) will be logged.

When an app is unfrozen, 
 [`ActivityManager system_server D  sync unfroze <pid> <package_name> for <reason>`](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/services/core/java/com/android/server/am/CachedAppOptimizer.java;l=1510) where `reason` is one of [`@UnfreezeReason`](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/proto_logging/stats/atoms.proto;l=14947), also defined in [`ActivityManagerInternal`](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/core/java/android/app/ActivityManagerInternal.java;l=347).


The app freezer can be controlled with the `cached_apps_freezer` setting in the `global` namespace. The value of `enabled` enables the app freezer, `disabled` disables it and `device_default` will let the system decide whether to enable the freezer or not, as [it may not be supported on all systems](https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/services/core/java/com/android/server/am/CachedAppOptimizer.java;l=1054) ([`cgroup freeze state](https://cs.android.com/android/kernel/superproject/+/common-android-mainline:common/Documentation/admin-guide/cgroup-v1/freezer-subsystem.rst) file needs to be present).

There is also a `use_freezer`  `device_config` setting in the  `activity_manager_native_boot` namespace that acts as a secondary setting.

https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/services/core/java/com/android/server/am/CachedAppOptimizer.java;l=1098

https://cs.android.com/android/platform/superproject/+/android-14.0.0_r9:frameworks/base/core/java/android/provider/Settings.java;l=17517
 
```shell
# To disable freezer
## adb
adb shell "settings put global cached_apps_freezer disabled"
## root
su -c "settings put global cached_apps_freezer disabled"

# To re-enable or restore freezer to default
## adb
adb shell "settings delete global cached_apps_freezer"
## root
su -c "settings delete global cached_apps_freezer"


# To check current settings value
## adb
adb shell "settings get global cached_apps_freezer"
## root
su -c "settings get global cached_apps_freezer"
```

To view current freezer status only.

```shell
## adb
adb shell "dumpsys activity | grep use_freezer="
## root
su -c "dumpsys activity | grep use_freezer="
```
Output:

```
  use_freezer=true
```

To view current freezer status, settings and view list of frozen processes.

```shell
## adb
adb shell "dumpsys activity | sed -n '/use_freezer=/,/CacheOomRanker/p'"
## root
su -c "dumpsys activity | sed -n '/use_freezer=/,/CacheOomRanker/p'"
```

Output:

```
use_freezer=true
  freeze_statsd_sample_rate=0.1
  freeze_debounce_timeout=10000
  freeze_exempt_inst_pkg=false
  freeze_binder_enabled=true
  freeze_binder_threshold=1000
  freeze_binder_divisor=4
  freeze_binder_offset=500
  Apps frozen: 102
   ...
    2803411: 14658 com.termux.api
```

Related commits:

- https://cs.android.com/android/_/android/platform/frameworks/base/+/79825c6f2f8b46808e4b431fe52b3be78f1e8ac8
- https://cs.android.com/android/_/android/platform/frameworks/base/+/8a2e9fbef822c0acce307a9552c7f2b4c6a0fc59
- https://cs.android.com/android/_/android/platform/frameworks/base/+/9dbe8e97f85cf0c505efc9615a15286e5886618e
- https://cs.android.com/android/_/android/platform/frameworks/base/+/9e82d90c49848bb55397d416cc0faaac1f1dfa2e
- https://cs.android.com/android/_/android/platform/frameworks/base/+/99111bf436fdbc84184713030ab5b7a98aae89f9

Disabling battery optimizations for the app does not prevent it from being frozen. Sending an intent to start a foreground service for the app with `am start` will not resume a frozen app, like [`termux-api-start`](https://github.com/termux/termux-api-package/blob/master/scripts/termux-api-start.in) does to start [`KeepAliveService`](https://github.com/termux/termux-api/blob/v0.50.1/app/src/main/java/com/termux/api/KeepAliveService.java), however if started in unfrozen, state, it does seem to prevent app being shortly frozen, I guess until it is killed.

The issue can be reproduced by removing Termux:API from recent apps screen after launching its activity, waiting a while and then running `termux-api` commands. If command hangs and `dumpsys` shows `com.termux.api` in frozen apps list or `freezing <pid> com.termux.api` was logged in system `logcat`, then start the Termux:API activity from launcher and command should resume. If you kill the hung command with `ctrl+c` before starting Termux:API, then when app resumes again, `ResultReturner` will throw a `java.io.IOException: Connection refused` exception since it will fail to connect to the socket previously started by `termux-api` that is no longer available.

**Comment by Daplugg23** [comment]
After installing pkg install root-repo and then apt update && apt upgrade. It's now working. Thank you for the help.

**Comment by agnostic-apollo** [comment]
Android 14 requires `termux-am` `v0.6.0`.

https://github.com/termux/TermuxAm/commit/8761289ba6ff3639e9b9c6af0d2d267e01dfccdb

Welcome.

**Comment by Daplugg23** [comment]
After performing a factory reset on my phone, I'm encountering the same issue again where the termux-usb -l command in Termux hangs. Despite attempting the same troubleshooting steps as before, the problem persists. Attached is the logcat output for your review. Could you please assist in resolving this?
[logcat.txt](https://github.com/termux/termux-api/files/13398566/logcat.txt)

**Comment by CrazyCraftix** [comment]
I have the same problem. For me, it seems to affect all termux-api commands.

`dumpsys` does not show `com.termux.api` as frozen and I have version `0.6.0` of the package `termux-am` installed.

I'm using the latest F-Droid versions of Termux and Termux API.
I don't want to try a newer, non-F-Droid build (as done in #640), because I'm afraid I might lose the keys inside the Android Keystore (I was using [tergent](https://github.com/aeolwyr/tergent)).

The only notable events between everything working and now (that I remember) was a minor system update and `pkg update && pkg upgrade`.

**Comment by txoof** [comment]
@alexfarseer 
Can you offer a more detailed description of how to work through these steps:

>  First update Termux to the GitHub build if you are using the F-Droid version. And get the latest build of Termux-API from here: https://github.com/termux/termux-api/actions/runs/6717239830

I just installed the December patch this morning (15 Dec) and all the termux-api commands hang as described above. I tried uninstalling and reinstalling the F-Droid package as well as `pkg uninstall termux-api && pkg install termux-api`

Current versions:
* termux-api pkg: 0.57
* termux-api fdroid 0.50.1

Thanks for the help!

**Comment by alexfarseer** [comment]
@txoof Okey, first of all uninstall the F-Droid packages. Latest build of Termux-API does not work with them. Get the Termux app from here: https://github.com/termux/termux-app/releases/tag/v0.118.0

And Termux-API app from here: https://github.com/termux/termux-api/actions/runs/6717239830

Install them both and give all the permissions to Termux-API, including the three special permissions from inside the app. After that run `termux-api-start` from Termux and all the API commands should work.

**Comment by txoof** [comment]
@alexfarseer 

Fantastic! Thank you for the instructions. This worked perfectly. 

When I run termux-api-start I get a notifciation: `Termux:API Error Error in ResultReturner`, but after that things work as expected.  I see the following on the command line:
```
$ termux-api-start                                                  
Starting service: Intent { cmp=com.termux.api/.KeepAliveService }
```

Any idea if this is important?

I have noticed that after a bit, the API commands hang again and I have to run the `termux-api-start` command again. Any solution for that other than shoving it into my scripts?

**Comment by txoof** [comment]
@agnostic-apollo 
I managed to get to a working version of termux:API by following alexfarseer's instructions above. Is there a more sustainable way to keep an up-to-date version of termux on my devices? Previously I was working from F-Droid and that worked great up until the December 2023 security patch broke something.

**Comment by alexfarseer** [comment]
@txoof Unfortunately it looks like `termux-api-start` is not working for you. There should be no errors and it should be alive until you stop it with `termux-api-stop`.

Run `termux-api-stop` after running `termux-api-start`, that will show you if the keep alive service was successfully started or not. Also, make sure you have enabled the special permissions inside Termux-API app. These permissions are battery optimisations and display over other apps.

**Comment by txoof** [comment]
Interesting. `termux-api-start` occasionally yields the notification error, but then all the API commands start functioning.

Whatever `api-start` does is only temporary and needs to be done again after an hour or so. I've solved this by sticking the command in all of the automation scripts that have API commands.

Thanks again for your help in getting this far.

**Comment by fritzbauer** [comment]
Same problem here after December update. Fixed with the termux & termux-api versions from github. As a newer termux-api version seems to resolve the issue, would it be possible to create a new release of termux-api so F-Droid builds the new version as well?

**Comment by astral-l** [comment]
Seconding the interest in a new release for F-Droid users. If a full release isn't planned soon, it might be easier to create a smaller patch release with just this issue fixed.

**Comment by agnostic-apollo** [comment]
Check https://github.com/termux/termux-api-package/commit/e8b001399ac679eb48590cfbc7d63e55104d8629 and https://github.com/termux/TermuxAm/commit/8a08e9bd31f3bc763f30ee1bf9c32204d0a89a53 for detail. Hanging issues for github action builds should be resolved now.

Update to `termux-api`  `v0.58.0` and `termux-am` `v0.7.0` with `apt install termux-api termux-am`. Packages are available in termux default repo and should be synced to other mirrors depending on their sync strategy.

I was wrong in the [above comment](https://github.com/termux/termux-api/issues/638#issuecomment-1813233924), a frozen can still be started with an intent. I did not previously realize that `termux-api` was hanging at the `read()` call instead of the process not starting again. You should not need to disable freezing as previously suggested, although it may cause some performance issues if app process gets frozen as wake up times should cause latency issues.

An app update is not required to fix this specific issue, but F-Droid builds are broken since exceptions are not caught and if app crash multiple times, it will never restart until manually started with a launcher and there is no launcher icon in fdroid builds. I should be able to start working on the pre-releases some time next week, it requires working on like 5 different repos to fix issues before I can make releases.

Those `ResultReturner` notifications can be opened to see the exception by tapping. There was likely due to the app failing to connect the socket sent by the `termux-api` command when it got unfroozen if `ctrl+c` was used to kill the command before that.

> `Caused by: java.io.InvalidClassException: com.termux.shared.models.ReportInfo; local class incompatible: stream classdesc serialVersionUID = -5165426368218339031, local class
 serialVersionUID = 3324698011185826218`
 
This is unrelated to current issue, probably some pre/post update mismatch issue.
 
 >I'm encountering the same issue again where the termux-usb -l command in Termux hangs. Despite attempting the same troubleshooting steps as before, the problem persists. Attached is the logcat output for your review. Could you please assist in resolving this?
> [logcat.txt](https://github.com/termux/termux-api/files/13398566/logcat.txt)

You are getting selinux denials, don't know why, maybe that is causing the hanging for some reason. Not sure why logs are labelled by `adb`, how are you running the `termux-usb` commads?
  
 > bash    : type=1400 audit(0.0:8641): avc:  granted  { execute } for  name="termux-usb" dev="dm-58" ino=86932 scontext=u:r:untrusted_app_27:s0:c90,c257,c512,c768 tcontext=u:object_r:app_data_file:s0:c90,c257,c512,c768 tclass=file app=com.termux.api
> adb     : type=1400 audit(0.0:8694): avc:  denied  { read } for  comm=64657669636520706F6C6C name="usb" dev="tmpfs" ino=494 scontext=u:r:untrusted_app_27:s0:c90,c257,c512,c768 tcontext=u:object_r:usb_device:s0 tclass=dir permissive=0 app=com.termux.api

**Comment by RoseDeSable** [comment]
Yesterday,
 I installed the termux:API from F-Droid app on my Samsung Tablet A8. However the commands of the termux-api package like termux-usb still hung. After a complete reboot of my tablet all the problems are solved.
 The operating system of my tablet always is on the newest software level.
